hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1725,6 +1729,173 @@ public void delete(final Path path, final boolean recursive,
+              ex.getErrorMessage(), ex);
+        }
+      }
+      LOG.debug(
+          String.format("Path %s is not a directory", path.toUri().getPath()));
+      deleteCheckOnParentBlob(path, tracingContext);
+      LOG.debug(String.format("Deleting path %s", path.toUri().getPath()));
+      client.deleteBlobPath(path, null, tracingContext);
+      LOG.debug(String.format("Path deleted %s", path.toUri().getPath()));
+    }
+  }
+
+  /**
+   * If parent blob is implicit directory and in case the blob deleted was the
+   * only blob in the directory, it will render path as non-existing. To prevent
+   * that happening, we check if the parent directory is implicit and if yes, we
+   * create marker-based directory.
+   *
+   * @param path path getting deleted
+   * @param tracingContext tracingContext to trace the API flow
+   * @throws IOException
+   */
+  private void deleteCheckOnParentBlob(final Path path,
+      final TracingContext tracingContext) throws IOException {
+    if (path.isRoot()) {
+      return;
+    }
+    Path parentPath = path.getParent();
+    if (parentPath.isRoot()) {
+      return;
+    }
+    LOG.debug(String.format("Checking parent of Path %s : %s exists",
+        path.toUri().getPath(), parentPath.toUri().getPath()));
+    try {
+      getBlobProperty(parentPath, tracingContext);
+      LOG.debug(
+          String.format("Parent of Path %s : %s exists", path.toUri().getPath(),
+              parentPath.toUri().getPath()));
+    } catch (AbfsRestOperationException ex) {
+      if (ex.getStatusCode() != HttpURLConnection.HTTP_NOT_FOUND) {
+        LOG.error(String.format("Checking parent of Path %s : %s failed",
+            path.toUri().getPath(), parentPath.toUri().getPath()), ex);
+        throw ex;
+      }
+      LOG.debug(String.format(
+          "Parent of Path %s : %s doesn't exist. Creating directory for the parentPath",
+          path.toUri().getPath(), parentPath.toUri().getPath()));
+      createDirectory(parentPath, null, FsPermission.getDirDefault(),
+          FsPermission.getUMask(
+              getAbfsConfiguration().getRawConfiguration()),
+          tracingContext);
+
+      LOG.debug(String.format("Directory for parent of Path %s : %s created",
+          path.toUri().getPath(), parentPath.toUri().getPath()));
+    }
+  }
+
+  private void deleteOnConsumedBlobs(final Path srcPath,
+      final ListBlobConsumer consumer,
+      final TracingContext tracingContext)
+      throws IOException {
+    ExecutorService deleteBlobExecutorService = Executors.newFixedThreadPool(
+        getAbfsConfiguration().getBlobDirDeleteMaxThread());
+    while (!consumer.isCompleted()) {
+      final List<BlobProperty> blobList = consumer.consume();
+      if (blobList == null) {
+        continue;
+      }
+      List<Future> futureList = new ArrayList<>();
+      for (BlobProperty blobProperty : blobList) {
+        futureList.add(deleteBlobExecutorService.submit(() -> {
+          try {
```

Comment:
blobProperty.getPath().toUri().getPath() -> same for this variable here

Reply:
done